### ISSUE TITLE
C API: Unexpected behaviour when multithreading calls of c_api's TF_SessionRun

### ISSUE THREAD

**Issue by Nucs** [tensorflow/tensorflow#32449]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): [Official download URL from TF's website. ](https://www.tensorflow.org/install/lang_c),Which is then downloaded via nuget package.
- TensorFlow version (use command below): Unable, we use c_api, the dll is downloaded from [here](https://storage.googleapis.com/tensorflow/libtensorflow/libtensorflow-cpu-windows-x86_64-1.14.0.zip)
- Python version: -
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: We use CPU
- GPU model and memory: -

**Describe the current behavior**
I'm working with our team on [Tensorflow.NET](https://github.com/SciSharp/TensorFlow.NET/) and we wanted to exploit C#'s multithreading capabilities with Tensorflow. 
Our library is a port of close to 1-1 from python to C# with couple of changes to allow multithreading to work.
We use a different Session and Graph for every Thread so every thread is actually isolated from the other Threads.
When calling `c_api.TF_SessionRun` from 10 parallel threads in an infinite loop for a minute or so - some sort of corruption occurs that causes Tensorflow to terminate with the following error code 

`Code -1073740791 (0xc0000409).` which is:
> Stack buffer overflow / overrun. Error can indicate a bug in the executed software that causes > stack overflow, leading to abnormal termination of the software.

This does not occur every run, in-fact it might occur 1/10 and only after 1-5 minutes of running.

In most cases, the program will just terminate but even more rarely the following message is printed:
> 2019-09-12 02:47:15.298642: F tensorflow/core/framework/tensor.cc:928] Check failed: buf_ null buf_ with non-zero shape size 15

We discussed the issue in our repository: https://github.com/SciSharp/TensorFlow.NET/issues/380

**Describe the expected behavior**
Run smoothly, `c_api.TF_SessionRun` expected to be thread-safe when used with a separate Session and Graph for every thread.

**Code to reproduce the issue**
https://github.com/Nucs/TensorFlowNetMultithreading
Requirements: Visual Studio 2017+
Happens both in Debug or Release builds.

The tensorflow.dll we use is not built on Debug, in order to use a custom tensorflow.dll with a .pdb, replace the one in `CalcEventsTFS\bin\Debug\net461 or netcoreapp2.2\` and then run `CalcEventsTFS.exe`

Simply build the solution and run the only project there is. 
I recommend to restart the program if it did not occur within 1 minute.

If you are getting `Unable to load DLL 'tensorflow': The specified module could not be found.`, 
Copy your own tensorflow.dll of version 1.40 CPU for windows to the output directory: 
> CalcEventsTFS\bin\Debug\net461 or netcoreapp2.2\

**Other info / logs**
We managed to get a dump after the termination of the program available from [here]( https://mega.nz/#!VVF3HKrS!r5sxOBsbccdVn93KpwS7EtxWSvtLlim2GwExDF8L9h4).

![image](https://user-images.githubusercontent.com/649919/64742960-c1d51c80-d506-11e9-9f2d-c4f7de2234b8.png)

**Comment by Nucs** [comment]
Any update on the issue? This bug prevents our users from using Tensorflow in a multithreaded manner in production.

**Comment by GeorgeS2019** [comment]
@Nucs Latest [tensorflow.dll ](https://github.com/tensorflow/tensorflow/issues/35405#issuecomment-653621878)for c# api

**Comment by roman-kruglov** [comment]
I experience similar symptoms. Is it at least reproduced? Confirmed? I guess it is not solved for sure.

We use the same graph and session, initialized once, to run inference with a call to `TF_SessionRun` from different threads. It is CPU inference in our case. Using more than one session adds a considerable amount or RAM usage (doubles, triples, etc - depending on the number of threads and sessions).

I thought it is safe to use like that, but now I'm not sure - can't even recall where I've seen that it was safe after all. The header file (namely c_api.h) sure doesn't say so, there is almost no mentions of thread safety at all.

Can somebody at least clarify if such usage is intended to work and is safe?

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.

**Comment by creativesalam** [comment]
> I experience similar symptoms. Is it at least reproduced? Confirmed? I guess it is not solved for sure.
> 
> We use the same graph and session, initialized once, to run inference with a call to `TF_SessionRun` from different threads. It is CPU inference in our case. The app is written in C++ and loads TF as a shared lib on Linux and Windows. Using more than one session adds a considerable amount or RAM usage (doubles, triples, etc - depending on the number of threads and sessions).
> 
> I thought it is safe to use like that, but now I'm not sure - can't even recall where I've seen that it was safe after all. The header file (namely c_api.h) sure doesn't say so, there is almost no mentions of thread safety at all.
> 
> Can somebody at least clarify if such usage is intended to work and is safe?
> 
> **P.S.** by the way, we used the latest TF version, way beyond 2.0. Doesn't help a bit.

Have you found any solution to that or you enforced some how to run it on single thread? I am facing the same problem. My priority is to use it in multi-thread environment.
